File: scripts/grid_view.py
#!env python3

from generate import Component, Property, init_method, Param, Method, Return

if __name__ == "__main__":
    w = Component(
        Type="appkit.GridView",
        super_type='appkit.View',
        description="a container that aligns views in a flexible grid of rows and columns",
        properties=[
            Property(name='numberOfRows', Type='int', readonly=True,
                     description='the number of rows in the grid view'),
            Property(name='numberOfColumns', Type='int', readonly=True,
                     description='the number of columns in the grid view'),
            Property(name='columnSpacing', Type='float64',
                     description='the column spacing for the grid view'),
            Property(name='rowSpacing', Type='float64',
                     description='the row spacing for the grid view'),
            Property(name='rowAlignment', Type='int', go_alias_type='GridRowAlignment',
                     description='the row alignment for the grid view'),
            Property(name='xPlacement', Type='int', go_alias_type='GridCellPlacement',
                     description='The placement of the cell within the grid column'),
            Property(name='yPlacement', Type='int', go_alias_type='GridCellPlacement',
                     description='The placement of the cell within the grid row'),

        ],
        methods=[
            init_method(
                name='initWithFrame',
                Type='appkit.GridView',
                params=[
                    Param(name='frameRect', Type='foundation.Rect'),
                ],
            ),
            # Method(
            #     name='gridViewWithViews',
            #     register_delegate=True,
            #     static=True,
            #     params=[Param(name='views', Type='appkit.View', array=True)],
            #     return_value=Return(Type='appkit.GridView'),
            #     description='creates a newly allocated grid view object with the specified array of arrays of views',
            # ),
            Method(
                name='gridViewWithNumberOfColumns',
                register_delegate=True,
                static=True,
                params=[Param(name='columnCount', Type='int'), Param(
                    name='rowCount', Type='int', objc_param_name='rows')],
                return_value=Return(Type='appkit.GridView'),
                description='creates a newly allocated grid view object with the specified number of columns and rows',
            ),
            Method(
                name='indexOfColumn',
                params=[Param(name='column', Type='appkit.GridColumn')],
                return_value=Return(Type='int'),
                description='returns the index of the specified grid column',
            ),
            Method(
                name='indexOfRow',
                params=[Param(name='row', Type='appkit.GridRow')],
                return_value=Return(Type='int'),
                description='returns the index of the specified grid row',
            ),
            Method(
                name='columnAtIndex',
                params=[Param(name='index', Type='int')],
                return_value=Return(Type='appkit.GridColumn'),
                description='returns the grid column object at the specified index',
            ),
            Method(
                name='rowAtIndex',
                params=[Param(name='index', Type='int')],
                return_value=Return(Type='appkit.GridRow'),
                description='returns the grid row object at the specified index',
            ),
            Method(
                name='addRowWithViews',
                params=[Param(name='views', Type='appkit.View', array=True)],
                return_value=Return('appkit.GridRow'),
                description='adds an array of views to a new row',
            ),
            Method(
                name='insertRowAtIndex',
                params=[
                    Param(name='index', Type='int'),
                    Param(name='views', Type='appkit.View', array=True, objc_param_name='withViews'),
                ],
                return_value=Return('appkit.GridRow'),
                description='inserts the array of view objects into the grid view at the index',
            ),
            Method(
                name='removeRowAtIndex',
                params=[
                    Param(name='index', Type='int'),
                ],
                description='removes the row from the grid view at the index',
            ),
            Method(
                name='moveRowAtIndex',
                params=[
                    Param(name='fromIndex', Type='int'),
                    Param(name='toIndex', Type='int', objc_param_name='toIndex'),
                ],
                description='moves the specified row to the new row location',
            ),
            Method(
                name='addColumnWithViews',
                params=[Param(name='views', Type='appkit.View', array=True)],
                return_value=Return('appkit.GridColumn'),
                description='adds an array of views to a new column',
            ),
            Method(
                name='insertColumnAtIndex',
                params=[
                    Param(name='index', Type='int'),
                    Param(name='views', Type='appkit.View', array=True, objc_param_name='withViews'),
                ],
                return_value=Return('appkit.GridColumn'),
                description='inserts the array of view objects into the grid view at the index',
            ),
            Method(
                name='removeColumnAtIndex',
                params=[
                    Param(name='index', Type='int'),
                ],
                description='removes the column from the grid view at the index',
            ),
            Method(
                name='moveColumnAtIndex',
                params=[
                    Param(name='fromIndex', Type='int'),
                    Param(name='toIndex', Type='int', objc_param_name='toIndex'),
                ],
                description='moves the specified column to the new column location',
            ),
            Method(
                name='cellAtColumnIndex',
                go_func_name='CellAt',
                params=[
                    Param(name='columnIndex', Type='int'),
                    Param(name='rowIndex', Type='int', objc_param_name='rowIndex'),
                ],
                return_value=Return('appkit.GridCell'),
                description='returns the grid cell object at the specified column and row index',
            ),
            Method(
                name='cellForView',
                params=[
                    Param(name='view', Type='appkit.View'),
                ],
                return_value=Return('appkit.GridCell'),
                description='returns the grid cell object that contains the given view or one of its ancestors',
            ),
            Method(
                name='mergeCellsInHorizontalRange',
                params=[
                    Param(name='hRange', Type='foundation.Range'),
                    Param(name='vRange', Type='foundation.Range', objc_param_name='verticalRange'),
                ],
                description='expands the cell at the top-leading corner of the horizontal and vertical range to cover the entire area',
            ),
            
        ]
    )
    w.generate_code()

File: appkit/grid_view.m
#import <AppKit/AppKit.h>
#import "grid_view.h"

long GridView_NumberOfRows(void* ptr) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView numberOfRows];
}

long GridView_NumberOfColumns(void* ptr) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView numberOfColumns];
}

double GridView_ColumnSpacing(void* ptr) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView columnSpacing];
}

void GridView_SetColumnSpacing(void* ptr, double columnSpacing) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView setColumnSpacing:columnSpacing];
}

double GridView_RowSpacing(void* ptr) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView rowSpacing];
}

void GridView_SetRowSpacing(void* ptr, double rowSpacing) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView setRowSpacing:rowSpacing];
}

long GridView_RowAlignment(void* ptr) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView rowAlignment];
}

void GridView_SetRowAlignment(void* ptr, long rowAlignment) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView setRowAlignment:rowAlignment];
}

long GridView_XPlacement(void* ptr) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView xPlacement];
}

void GridView_SetXPlacement(void* ptr, long xPlacement) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView setXPlacement:xPlacement];
}

long GridView_YPlacement(void* ptr) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView yPlacement];
}

void GridView_SetYPlacement(void* ptr, long yPlacement) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView setYPlacement:yPlacement];
}

void* GridView_NewGridView(NSRect frameRect) {
	NSGridView* gridView = [NSGridView alloc];
	return [[gridView initWithFrame:frameRect] autorelease];
}

void* GridView_GridViewWithNumberOfColumns(long columnCount, long rowCount) {
	return [NSGridView gridViewWithNumberOfColumns:columnCount rows:rowCount];
}

long GridView_IndexOfColumn(void* ptr, void* column) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView indexOfColumn:(NSGridColumn*)column];
}

long GridView_IndexOfRow(void* ptr, void* row) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView indexOfRow:(NSGridRow*)row];
}

void* GridView_ColumnAtIndex(void* ptr, long index) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView columnAtIndex:index];
}

void* GridView_RowAtIndex(void* ptr, long index) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView rowAtIndex:index];
}

void* GridView_AddRowWithViews(void* ptr, Array views) {
	NSGridView* gridView = (NSGridView*)ptr;
    NSMutableArray* objcViews = [[NSMutableArray alloc] init];
    void** viewsData = (void**)views.data;
    for (int i = 0; i < views.len; i++) {
    	[objcViews addObject:(NSView*)viewsData[i]];
    }
	return [gridView addRowWithViews:objcViews];
}

void* GridView_InsertRowAtIndex(void* ptr, long index, Array views) {
	NSGridView* gridView = (NSGridView*)ptr;
    NSMutableArray* objcViews = [[NSMutableArray alloc] init];
    void** viewsData = (void**)views.data;
    for (int i = 0; i < views.len; i++) {
    	[objcViews addObject:(NSView*)viewsData[i]];
    }
	return [gridView insertRowAtIndex:index withViews:objcViews];
}

void GridView_RemoveRowAtIndex(void* ptr, long index) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView removeRowAtIndex:index];
}

void GridView_MoveRowAtIndex(void* ptr, long fromIndex, long toIndex) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView moveRowAtIndex:fromIndex toIndex:toIndex];
}

void* GridView_AddColumnWithViews(void* ptr, Array views) {
	NSGridView* gridView = (NSGridView*)ptr;
    NSMutableArray* objcViews = [[NSMutableArray alloc] init];
    void** viewsData = (void**)views.data;
    for (int i = 0; i < views.len; i++) {
    	[objcViews addObject:(NSView*)viewsData[i]];
    }
	return [gridView addColumnWithViews:objcViews];
}

void* GridView_InsertColumnAtIndex(void* ptr, long index, Array views) {
	NSGridView* gridView = (NSGridView*)ptr;
    NSMutableArray* objcViews = [[NSMutableArray alloc] init];
    void** viewsData = (void**)views.data;
    for (int i = 0; i < views.len; i++) {
    	[objcViews addObject:(NSView*)viewsData[i]];
    }
	return [gridView insertColumnAtIndex:index withViews:objcViews];
}

void GridView_RemoveColumnAtIndex(void* ptr, long index) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView removeColumnAtIndex:index];
}

void GridView_MoveColumnAtIndex(void* ptr, long fromIndex, long toIndex) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView moveColumnAtIndex:fromIndex toIndex:toIndex];
}

void* GridView_CellAt(void* ptr, long columnIndex, long rowIndex) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView cellAtColumnIndex:columnIndex rowIndex:rowIndex];
}

void* GridView_CellForView(void* ptr, void* view) {
	NSGridView* gridView = (NSGridView*)ptr;
	return [gridView cellForView:(NSView*)view];
}

void GridView_MergeCellsInHorizontalRange(void* ptr, NSRange hRange, NSRange vRange) {
	NSGridView* gridView = (NSGridView*)ptr;
	[gridView mergeCellsInHorizontalRange:hRange verticalRange:vRange];
}
